SPEC.md

```diff
@@ -0,0 +1,258 @@
+```
+
+To illustrate composing fetchers, a polling fetcher can send a query operation to an http fetcher on a specified interval. Each request is routed through the http fetcher and returned through the observable from the polling fetcher. Using this type of delegation, fetchers of discrete functionality can form stacks.
+
+<p align="center">
+  <br>
+  <img src="images/polling-stack.png" alt="Polling Fetcher Stack"/>
+</p>
+
+
+```js
+const pollingFetcher = new PollingFetcher(
+  10000, //Polling Interval in ms
+  new HTTPFetcher({ uri: 'http://api.githunt.com/graphql' })
+);
+```
+
+In addition to the query, operationName, and variables, `Operation` includes a context, which is modified and passed down the stack of fetchers. In addition, this context is sent to the server to through the query body. In this example, `context` is used by a polling fetcher to tell a caching fetcher whether a request should be returned from the cache or the network.
+
+<p align="center">
+  <br>
+  <img src="images/context.png" alt="Context Passing Stack Example"/>
+</p>
+
+## API
+
+The fetcher interface contains a single method:
+
+```js
+interface ApolloFetcher{
+  request(Operation) : Observable<FetchResult>
+}
+```
+
+An `Operation` contains all the information necessary to execute a GraphQL query, including the GraphQL AST:
+
+```js
+Operation {
+  query: DocumentNode,
+  operationName: string,
+  variables: object,
+  context: object,
+}
+```
+
+An `Observable` is returned from `request` and notifies a `Subscriber` on the life-cycle of a GraphQL request:
+
+```js
+Observable{
+  //Causes the underlying producer to start
+  subscribe(Subscriber<FetchResult>) : UnsubscribeHandler
+  subscribe(
+    next(FetcheResult) => void,
+    error(any) => void,
+    complete() => void) : UnsubscribeHandler
+}
+```
+
+A `Subscriber` is passed to `subscribe`
+
+```js
+Subscriber<T>{
+  next(T),
+  error(any),
+  complete(),
+}
+```
+
+`UnsubscribeHandler` is returned from `subscribe`
+
+```js
+type UnsubscribeHandler = () => void
+```
+
+`FetchResult` is passed to the `next` callback of `Subscriber`
+
+```js
+FetchResult{
+  data?: object,
+  errors?: object[],
+  extensions?: any,
+  context?:object,
+}
+```
+
+## More examples
+
+Here is a list of planned fetchers with selected diagrams:
+
+Base:
+
+* Websocket: subscriptions support
+* Batching-\*: support for batching operations with a transport
+
+<p align="center">
+  <br>
+  <img src="images/batch-fetcher.png" alt="Batch Fetcher"/>
+</p>
+
+Intermediate:
+
+* Polling: repeats requests on a specified interval
+
+<p align="center">
+  <br>
+  <img src="images/polling-fetcher.png" alt="Polling Fetcher"/>
+</p>
+
+* Caching: returns data if result in the cache and stores data in cache on response
+* Compose: combines a list of fetchers (would require additional semantics around constructor argument order)
+* Dedup: saves query signatures that are currently on the wire and returns the result for all of those queries
```

Comment:
Things like `Dedup`, `Retry`, and one I'd like to see (`Catch`) don't make sense as fetchers to me, but rather middle / afterwares which run the fetcher call stack in a modified way. They aren't transport specific (i.e where the data comes from) but are operational (i.e. how to call the request). I'd rather see a `Fetcher` as a way to get data and it have the ability for its control flow to be modified by other utilities. This would make things like logging and analytics easy to integrate as well since they are essentially side effects.